design-documents/service-isolation/inter-service-communication.md

```diff
@@ -0,0 +1,294 @@
+ 
+The way exceptions thrown by services are serialized right now for RESTful web API needs to be changed for inter-service endpoints.
+Exceptions are serialized without their type information provided - this info is crutial for services to communicate with each other so
+the type information will be added to the output.
+ 
+_inter_webapi.xml_ files will be contained in separate <ModuleName>InterApi modules and will have to be installed
+on nodes accepting remote inter-service requests.
+ 
+##### Changes for retries mechanism
+The invoker implementation based on RESTful web APIs will offer default retries mechanism, but developers will be able to disable it
+by using _di.xml_ (see Invoker API and SPI [here](invoker.md)) for when they would want to delegate this responsibility to a service mash.
+ 
+When default retries mechanism is enabled _X-Magento-Call-Id_ header will be added to inter-service requests to identify calls.
+It will be randomly generated. Then if a timeout reached while waiting for a service to respond the invoker will send the same request
+again configured number of times with the same call ID. The node receiving requests will store the call ID in cache as a key and
+assign an empty value to it. When the node will have the service call result it will store it with the call ID. That way when a
+duplicate request comes in if the node can't find the call ID in cache it will execute the requested service method as usual,
+if the cache has the call ID that would mean the service method is being executed and we'll have to wait until the original process
+finishes it and writes the response to cache, if both call ID and the result are in cache then we just return the result. That way
+we will avoid executing duplicate calls when retries mechanism is applied.
+ 
+The timeout and tries limit will be configured via di.xml (see the invoker API and SPI [here](invoker.md)).
+    
+#### Service mashes and balancers
+Since we going to contact services via HTTP it is possible for configured service base URLs to actually lead to service mashes or
+balancers - that way the retries and lookup mechanism may be delegated.
+
+#### MQ RPC for gateway
+Magento has another way to remotely call services - RPC via a message queue. The problem with it is that right now it can only work
+with strings as arguments for service contracts and daemons written in PHP to process queues will show slower results in handling
+multiple requests than web servers like Nginx or Apache when using RESTful web API for gateway.
+
+#### Front node
+Front node is a Magento installation with Framework and Webapi/GraphQl modules.
+It will handle actual authentication and authorization of incoming requests and pass the auth info to remote services
+as well as accepting requests from clients (frontend, 3rd party systems).
+ 
+This is the node that will be available on a public network.
+
+#### Flow
+Entities:
+* Front node
+  * Modules installed:
```

Comment:
Putting all code in one module to enable remote communication for a module makes it easier to develop but has deployment complications. As there could be changes that are relevant to multiple services, changing module for one service will require redeploying other services. Need to think if we can make it more granular to avoid this issue.

Reply:
if a service adds a remote API (endpoint) that will not affect other services communicating with it using an older declarations